### PR TITLE
C2901 & C2902: New checks for unnecessary lambda expression usage

### LINKED ISSUE #5976: New lint: Use `def` instead of lambda expression for bare assignments and calls
### Current problem

Currently these uses of lambda expressions don't raise any lint.
A)
```python
a = lambda x: x
```
(^ Same as E731 in flake8)

B)
```python
(lambda x: x)("hello world")
```
(not flagged by flake8 or pylint)

### Desired solution

A)
```python
a = lambda x: x
```
raises lint to tell user to create a normal function instead of assigning a lambda.
Should be replaced with
```python
def a(x)
    return x
```

B)
```python
y = (lambda x: x.upper())("hello world")
```
raises lint telling user to just execute the lambda body since there's no need to define a function or lambda expression to compute something once.
Should be replaced with
```python
x = "hello world"
y = x.upper()
```
### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jpy-git**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Closes #5976.

Adds new check `unnecessary-lambda-assignment`: `Lambda expression assigned to a variable. Define a function using the "def" keyword instead.`

From [PEP8](https://peps.python.org/pep-0008/):
```
Always use a def statement instead of an assignment statement that binds a lambda expression directly to an identifier:
# Correct:
def f(x): return 2*x
# Wrong:
f = lambda x: 2*x
```

This is equivalent to `E731` in flake8 but additionally addresses several false negatives that are missed by that implementation: https://github.com/PyCQA/pycodestyle/issues/1061

Also adds a new check `unnecessary-direct-lambda-call`: `Lambda expression called directly. Execute the expression inline instead.`
i.e. `a**2` not `(lambda x: x**2)(a)`
This prevents a weird redundant usage of lambda functions that I've seen some junior devs do before.

### CODE REVIEW

File: tests/functional/u/unnecessary/unnecessary_lambda_assignment.py
Comment (jpy-git) [inline tests/functional/u/unnecessary/unnecessary_lambda_assignment.py:30]: Will uncomment this test case once #5999 gets merged in

**Comment by jpy-git** [comment]
There were a lot of existing tests using that I had to add ignores to in that last commit, so look at the commit before that for easier reviewing

File: pylint/checkers/lambda_expressions.py
Comment (jacobtylerwalls) [inline pylint/checkers/lambda_expressions.py:36]:
```suggestion
            "rather than executing its contents inline.",
```

File: pylint/checkers/lambda_expressions.py
Comment (jacobtylerwalls) [inline pylint/checkers/lambda_expressions.py:54]:
```suggestion
            # see if any lambdas are assigned to a variable.
```

**Review by jacobtylerwalls — COMMENTED** [review]
This is hilariously unlikely to happen, but what do you think about attempting to catch this case? (It doesn't emit a warning even if you remove the disable):
https://github.com/PyCQA/pylint/blob/edf753e6864224ac0144c0ac12c8d89baa93e097/tests/functional/a/assignment/assignment_expression.py#L44

**Comment by jpy-git** [comment]
@jacobtylerwalls I've extended the check to flag assignment via named expressions as well 👍 
Currently the tests for python<=3.7 are failing because walrus operators were only introduced in 3.8. Is there a way to only run some lines/files for certain versions?

**Comment by Pierre-Sassoulas** [comment]
> Currently the tests for python<=3.7 are failing because walrus operators were only introduced in 3.8. Is there a way to only run some lines/files for certain versions?

Yes, look at a .rc files in `tests/functional/`` there should be examples of use for ``min_version``

**Comment by jpy-git** [comment]
Thanks!

**Review by jacobtylerwalls — APPROVED** [review]

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Not sure if this is an option, but as a collaborator I can re-run actions. Can we allow this for contributors as well? That might make this more workable.

File: tests/functional/l/lambda_use_before_assign.py
Comment (DanielNoord) [inline tests/functional/l/lambda_use_before_assign.py:3]:
```suggestion

from __future__ import print_function
```

File: tests/functional/u/undefined/undefined_variable.py
Comment (DanielNoord) [inline tests/functional/u/undefined/undefined_variable.py:4]:
```suggestion
# pylint: disable=unnecessary-lambda-assignment

```

File: tests/functional/u/unnecessary/unnecessary_direct_lambda_call.py
Comment (DanielNoord) [inline tests/functional/u/unnecessary/unnecessary_direct_lambda_call.py:3]:
```suggestion

y = (lambda x: x**2 + 2*x + 1)(a)  # [unnecessary-direct-lambda-call]
```

File: tests/functional/u/unnecessary/unnecessary_lambda_assignment.py
Comment (DanielNoord) [inline tests/functional/u/unnecessary/unnecessary_lambda_assignment.py:3]:
```suggestion

a = lambda x: x  # [unnecessary-lambda-assignment]
```

File: tests/functional/u/unnecessary/unnecessary_lambda_assignment.py
Comment (DanielNoord) [inline tests/functional/u/unnecessary/unnecessary_lambda_assignment.py:1]: Could you add the following test case:
```python
DICT = {1: lambda x: x, 2: lambda x: x + 1}

for key, value in DICT.items():
    print(value(key))
```

Similar patterns are reasons why I use the `lambda x: x` pattern and I want to make sure we don't accidentally regress and flag this in the future.

**Review by DanielNoord — CHANGES_REQUESTED** [review]

**Comment by Pierre-Sassoulas** [comment]
> Can we allow this for contributors as well?

I don't think so, but the trick to open/close the issue should do most of the time.

**Comment by jpy-git** [comment]
is there a reason why we don't increase the `timeout-minutes` settings in the yaml file?

**Comment by Pierre-Sassoulas** [comment]
From what I understand the pipeline that are stuck are stuck because the runner used by github has only one core and we try to use two. So it will fail even if we wait one hour. 15mn permit to free the worker sooner and have overall better availability.

File: tests/functional/u/unnecessary/unnecessary_lambda_assignment.py
Comment (jpy-git) [reply tests/functional/u/unnecessary/unnecessary_lambda_assignment.py:1]: Yeah this is fine, I've added it in to the unit test 😄

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM, but I think it would be a nice check to inaugurate the false positive github action primer when it's available. It's a default checker and it's going to be run in a lot of code once it lands in pylint.

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas @jpy-git For some reason the latest merge of `main` broke this. Could either of you look into this?

After that, I think we can go ahead and merge this as we shouldn't let this be blocked by the eventual update to the `primer` imo.

**Comment by Pierre-Sassoulas** [comment]
Sure let me fix this, I should have merged that a long time ago. I think the issue is only with the checker number that has been claimed by another new checker since.

**Comment by Pierre-Sassoulas** [comment]
pypy has column/line variation compared to cpython. This is something we do not control and we're probably never going to change the parsing to have the right column in both. Should we still take the column/line info for under PYPY ?

**Comment by DanielNoord** [comment]
> @DanielNoord pypy has column/line variation compared to cpython. This is something we do not control and we're probably never going to change the parsing to have the right column in both. Should we still take the column/line into account under PYPY ?

Let's ignore this test on `PyPy`. This is a very rare (and only) case where the `col_offset` (and not `end_col_offset`) differ. Actually I think the column offset of `PyPy` is more "correct", but that's just my opinion :smile. I don't think we should stop checking `col_offset` on `PyPy` all together as this is the only case where it differs.

**Comment by DanielNoord** [comment]
Thanks @jpy-git for contributing the new checker and thanks to all reviewers 🎉

**Pull request merged**
